javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/muzzle/collector/ReferenceCollector.java

```diff
@@ -201,6 +203,89 @@ private void collectHelperClasses(
     return references;
   }
 
+  public void prune() {
+    // helper classes that may help another helper class implement an abstract library method
+    // must be retained
+    // for example if helper class A extends helper class B, and A also implements a library
+    // interface L, then B needs to be retained so that it can be used at runtime to verify that A
+    // implements all of L's methods.
+    // Super types of A that are not also helper classes do not need to be retained because they can
+    // be looked up on the classpath at runtime, see HelperReferenceWrapper.create().
+    Set<Reference> helperClassesParticipatingInLibrarySuperType =
+        getHelperClassesParticipatingInLibrarySuperType();
+
+    for (Iterator<Reference> i = references.values().iterator(); i.hasNext(); ) {
+      Reference reference = i.next();
+      if (instrumentationClassPredicate.isProvidedByLibrary(reference.getClassName())) {
+        // these are the references to library classes which need to be checked at runtime
+        continue;
+      }
+      if (helperClassesParticipatingInLibrarySuperType.contains(reference)) {
+        // these need to be kept in order to check that abstract methods are implemented,
+        // and to check that declared super class fields are present
+        //
+        // can at least prune the constructor, since that cannot be used to help implement an
+        // abstract library method
+        reference.getMethods().removeIf(method -> method.getName().equals("<init>"));
```

Comment:
Hmm, this is already checked in `ReferenceCollectingClassVisitor#visitMethod()`, but I suppose that does not detect all possible usages (helper classes can create other helper classes); should we move that check here?

Reply:
good idea, done